File: fixtures/sample-repos/selenium/pages/actions/login_actions_bad.py
"""VIOLATION FILE: expect-in-action + selector-leak.

This action class demonstrates two rules:
- expect-in-action: `assert` calls belong in spec files, not action classes.
- selector-leak: raw `querySelector(...)` chains belong in locator classes,
  not action classes.

The canonical pattern lives in login_actions.py; this file exists to
demonstrate the rules.
"""
from selenium.webdriver.remote.webdriver import WebDriver

from pages.locators.login_locators import LoginLocators


class LoginActionsBad:
    def __init__(self, driver: WebDriver):
        self.locators = LoginLocators(driver)

    def sign_in(self, email: str, password: str) -> None:
        # expect-in-action — assertion in an action class.
        assert (isinstance(email, str))
        assert (isinstance(password, str))

        # selector-leak — raw querySelector chain in an action/page class.
        self.locators.driver.execute_script(
            "const e = document.querySelector('[role=\"textbox\"][aria-label=\"Email\"]');"
            "e.value = arguments[0];",
            email,
        )
        self.locators.driver.execute_script(
            "const p = document.querySelector('[role=\"textbox\"][aria-label=\"Password\"]');"
            "p.value = arguments[0];",
            password,
        )
        self.locators.visible(self.locators.SIGN_IN).click()

File: fixtures/sample-repos/webdriverio/test/pageobjects/actions/LoginActionsBad.ts
// VIOLATION FILE: expect-in-action + selector-leak.
//
// This action class demonstrates two rules:
// - expect-in-action: `expect` calls belong in spec files, not action classes.
// - selector-leak: raw selector chains belong in locator classes, not action
//   classes. The spec also demonstrates WebdriverIO's `$('...')` shorthand
//   which the scanner now detects.
//
// The canonical pattern lives in LoginActions.ts; this file exists to
// demonstrate the rules that Gavel's self-check reports.
import { LoginLocators } from '../locators/LoginLocators';

export class LoginActionsBad {
  private locators = new LoginLocators();

  async signIn(email: string, password: string): Promise<void> {
    // expect-in-action: an action class must not call `expect`. The
    // spec is the assertion layer.
    expect(email).toBeTruthy();
    expect(password).toBeTruthy();

    // selector-leak: raw querySelector chain inside the action class. Move
    // the selector into LoginLocators and reference it by name.
    await browser.execute(
      (selector) => document.querySelector(selector),
      '[role="textbox"][aria-label="Email"]',
    );
    await browser.execute(
      (selector) => document.querySelector(selector),
      '[role="textbox"][aria-label="Password"]',
    );
    await this.locators.signInButton.click();
  }
}
